software/hdmi2pcie/kernel/main.c

```diff
@@ -0,0 +1,744 @@
+	.vidioc_s_fmt_vid_cap = hdmi2pcie_s_fmt_vid,
+	.vidioc_g_fmt_vid_cap = hdmi2pcie_g_fmt_vid,
+	.vidioc_enum_fmt_vid_cap = hdmi2pcie_enum_fmt_vid,
+
+	.vidioc_s_dv_timings = hdmi2pcie_s_dv_timings,
+	.vidioc_g_dv_timings = hdmi2pcie_g_dv_timings,
+	.vidioc_enum_dv_timings = hdmi2pcie_enum_dv_timings,
+	.vidioc_query_dv_timings = hdmi2pcie_query_dv_timings,
+	.vidioc_dv_timings_cap = hdmi2pcie_dv_timings_cap,
+
+	.vidioc_enum_input = hdmi2pcie_enum_input,
+	.vidioc_g_input = hdmi2pcie_g_input,
+	.vidioc_s_input = hdmi2pcie_s_input,
+
+	.vidioc_reqbufs = vb2_ioctl_reqbufs,
+	.vidioc_create_bufs = vb2_ioctl_create_bufs,
+	.vidioc_querybuf = vb2_ioctl_querybuf,
+	.vidioc_qbuf = vb2_ioctl_qbuf,
+	.vidioc_dqbuf = vb2_ioctl_dqbuf,
+	.vidioc_expbuf = vb2_ioctl_expbuf,
+	.vidioc_streamon = vb2_ioctl_streamon,
+	.vidioc_streamoff = vb2_ioctl_streamoff,
+
+	.vidioc_log_status = v4l2_ctrl_log_status,
+	.vidioc_subscribe_event = v4l2_ctrl_subscribe_event,
+	.vidioc_unsubscribe_event = v4l2_event_unsubscribe,
+};
+
+static const struct v4l2_ioctl_ops hdmi2pcie_ioctl_out_ops = {
+	.vidioc_querycap = hdmi2pcie_querycap,
+
+	.vidioc_try_fmt_vid_out = hdmi2pcie_try_fmt_vid,
+	.vidioc_s_fmt_vid_out = hdmi2pcie_s_fmt_vid,
+	.vidioc_g_fmt_vid_out = hdmi2pcie_g_fmt_vid,
+	.vidioc_enum_fmt_vid_out = hdmi2pcie_enum_fmt_vid,
+
+	.vidioc_s_dv_timings = hdmi2pcie_s_dv_timings,
+	.vidioc_g_dv_timings = hdmi2pcie_g_dv_timings,
+	.vidioc_enum_dv_timings = hdmi2pcie_enum_dv_timings,
+	.vidioc_query_dv_timings = hdmi2pcie_query_dv_timings,
+	.vidioc_dv_timings_cap = hdmi2pcie_dv_timings_cap,
+
+	.vidioc_enum_output = hdmi2pcie_enum_output,
+	.vidioc_g_output = hdmi2pcie_g_output,
+	.vidioc_s_output = hdmi2pcie_s_output,
+
+	.vidioc_reqbufs = vb2_ioctl_reqbufs,
+	.vidioc_create_bufs = vb2_ioctl_create_bufs,
+	.vidioc_querybuf = vb2_ioctl_querybuf,
+	.vidioc_qbuf = vb2_ioctl_qbuf,
+	.vidioc_dqbuf = vb2_ioctl_dqbuf,
+	.vidioc_expbuf = vb2_ioctl_expbuf,
+	.vidioc_streamon = vb2_ioctl_streamon,
+	.vidioc_streamoff = vb2_ioctl_streamoff,
+
+	.vidioc_log_status = v4l2_ctrl_log_status,
+	.vidioc_subscribe_event = v4l2_ctrl_subscribe_event,
+	.vidioc_unsubscribe_event = v4l2_event_unsubscribe,
+};
+
+static const struct v4l2_file_operations hdmi2pcie_fops = {
+	.owner = THIS_MODULE,
+	.open = v4l2_fh_open,
+	.release = vb2_fop_release,
+	.unlocked_ioctl = video_ioctl2,
+	.read = vb2_fop_read,
+	.mmap = vb2_fop_mmap,
+	.poll = vb2_fop_poll,
+};
+
+static int hdmi2pcie_register_video_dev(struct pci_dev *pdev, struct vid_channel *chan, uint8_t dir)
+{
+	static const struct v4l2_dv_timings timings_def = V4L2_DV_BT_CEA_1920X1080P60;
+	struct video_device *vdev = &chan->vdev;
+	struct vb2_queue *q = &chan->queue;
+	int ret;
+
+	chan->dir = dir;
+
+	ret = v4l2_device_register(&pdev->dev, &chan->v4l2_dev);
+	if (ret) {
+		dev_err(&pdev->dev, "v4l2_device_register failed, ret=%d\n", ret);
+		return ret;
+	}
+
+	// Point to last 4 bytes of a 16MiB area
```

Comment:
I think you want to say something like "The last 4 bytes contain the the frame pointer buffer index. This is the index which the FPGA firmware is currently writing to".

Reply:
I've updated the comment to say more about what this variable is used for